designs/0003-parallel_autodiff.md

```diff
@@ -0,0 +1,705 @@
+- Feature Name: parallel_ad
+- Start Date: 2019-04-20
+- RFC PR: (leave this empty)
+- Stan Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The goal of this RFC is to expand the CPU based parallelism
+capabilities of Stan math. This RFC proposes to introduce a new
+parallel map with automatic sharding and a parallel reduce which
+combines automatic sharding with vectorization.
+
+
+
+# Motivation
+[motivation]: #motivation
+
+Currently `map_rect` allows map parallelism, but it brings a few
+restrictions:
+
+1. Very user unfriendly due to the requirement to pack and unpack data
+   and parameters passed into the function.
+2. The user must provide a sensible sharding block size. No work load
+   balancing is done during execution.
+3. Vectorization must be traded for parallelism in most applications.
+
+In short, we currently make it very hard for users to use parallelism
+and moreover the current facility cannot speed up problems where we
+have to give up vectorization in exchange. 
+
+This RFC intends to make parallelism available to a much broader user
+base in a user-friendly way (addressing point 1+2) and that
+parallelism will scale much better as vectorization is combined with
+parallelism in a coherent approach (point 3).
+
+So ideally we can get more users in need for parallelism to actually
+use it and I also expect that vectorized parallelism will greatly
+enhance the problem set where parallelism actually speeds up a Stan
+program. For example, `map_rect` right now is not able to speed up
+programs which can take advantage of vectorization, but still have to
+process medium to large data sets. A parallel reduce with automatic
+sharding and vectorization will instead allow to speed up such
+applications in full automation. Therefore, the cpu resources
+allocated to a Stan chain can be scaled as needed (nowadays one can
+get up to 96core machines on clusters or the cloud like AWS).
+
+Clearly large Stan runs which take multiple hours will benefit from
+the proposed parallelism, but very likely even smaller problems can be
+speed up whenever sufficient CPU cores are available.
+
+Use cases where this RFC will add value by speeding up execution are:
+
+- Parallel accumulation of independent log-likelihood terms with very
+  good scalability. The good scalability is achieved by combining
+  automatic sharding with vectorization.
+- Parallel evaluation of map where each iteration is independent.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The proposed signatures for the parallel map/reduce with automatic
```

Comment:
It would be really useful to include:
1. a `map_rect` example in the Stan language
2. the proposed `parallel_map` example in the Stan language

Then move the C++ lower as a secondary focus. It's how we implement it, not design.